execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs

```diff
@@ -4024,3 +4026,339 @@ fn should_distribute_delegation_rate_full_after_upgrading() {
+    let mut next_era = builder.get_era() + 1;
+
+    // first step after funding, adding bid and delegating.
+    let step_request = StepRequestBuilder::new()
+        .with_parent_state_hash(builder.get_post_state_hash())
+        .with_protocol_version(ProtocolVersion::V1_0_0)
+        .with_next_era_id(next_era)
+        .build();
+
+    builder.step(step_request).expect("should step");
+
+    let delegator = get_delegator_bid(&mut builder, VALIDATOR_1.clone(), DELEGATOR_1.clone());
+
+    assert!(delegator.is_some());
+    assert_eq!(
+        delegator.unwrap().staked_amount().to_owned(),
+        U512::from(DELEGATOR_1_STAKE)
+    );
+
+    next_era += 1;
+
+    let step_request = StepRequestBuilder::new()
+        .with_parent_state_hash(builder.get_post_state_hash())
+        .with_protocol_version(ProtocolVersion::V1_0_0)
+        .with_next_era_id(next_era)
+        .build();
+
+    builder.step(step_request).expect("should step");
+
+    // undelegate in the era right after we delegated.
+    undelegate(
+        &mut builder,
+        *DELEGATOR_1_ADDR,
+        DELEGATOR_1.clone(),
+        VALIDATOR_1.clone(),
+        U512::from(DELEGATOR_1_STAKE),
+    );
+    let delegator = get_delegator_bid(&mut builder, VALIDATOR_1.clone(), DELEGATOR_1.clone());
+    assert!(delegator.is_none());
+
+    let withdraws = builder.get_withdraws();
+    let unbonding_purses = withdraws
+        .get(&VALIDATOR_1_ADDR)
+        .expect("should have validator entry");
+    let delegator_unbond_amount = unbonding_purses
+        .iter()
+        .find(|up| *up.unbonder_public_key() == DELEGATOR_1.clone())
+        .expect("should be unbonding purse");
+
+    assert_eq!(
+        *delegator_unbond_amount.amount(),
+        U512::from(DELEGATOR_1_STAKE),
+        "unbond purse amount should match staked amount"
+    );
+
+    // step until validator receives rewards.
+    for _ in 0..2 {
+        let step_request = StepRequestBuilder::new()
+            .with_parent_state_hash(builder.get_post_state_hash())
+            .with_protocol_version(ProtocolVersion::V1_0_0)
+            .with_next_era_id(builder.get_era() + 1)
+            .build();
+
+        builder.step(step_request).expect("should step");
+    }
+
+    // validator receives rewards after this step.
+    let step_request = StepRequestBuilder::new()
+        .with_parent_state_hash(builder.get_post_state_hash())
+        .with_protocol_version(ProtocolVersion::V1_0_0)
+        .with_reward_item(RewardItem::new(VALIDATOR_1.clone(), BLOCK_REWARD))
+        .with_next_era_id(builder.get_era() + 1)
+        .build();
+
+    builder.step(step_request).expect("should step");
+
+    // Delegator should not remain delegated even though they were eligible for rewards in the
+    // second era.
+    let delegator = get_delegator_bid(&mut builder, VALIDATOR_1.clone(), DELEGATOR_1.clone());
+    assert!(delegator.is_none());
+}
+
+#[test]
+fn delegator_full_unbond_during_first_reward_era() {
```

Comment:
There's only a very slight difference between these two test cases.  It'd be good to briefly document the two tests, highlighting the difference between them.